README.md
# MapWrap :pushpin:

[![npm version](https://badge.fury.io/js/mapwrap.svg)](https://badge.fury.io/js/mapwrap) ![npm bundle size](https://img.shields.io/bundlephobia/min/mapwrap.svg)

### [Changelog - updated for 2.1.0](https://github.com/pixeltopic/mapwrap/blob/master/CHANGELOG.md)

## About

MapWrap is a simple wrapper around the Google Maps API for Geocoding, Directions, and Places (Nearby Search, Place Details). It provides simple methods to access data members in the API response with error checking.

MapWrap also implements a simple LRU cache to speed up requests and decrease the amount of unnecessary calls to the server. It is fully compliant with Google's restriction of not caching data for over 30 days.

## Installation
```
npm install --save mapwrap
```

## Usage

```js
const MapWrap = require("mapwrap");

const mapwrap = MapWrap({
  // DEFAULT_API_KEY is all you need to get started. 
  // All other config properties are optional.
  DEFAULT_API_KEY: "my_google_api_key",
  // Optional keys with priority over default for specific API services.
  useRestrictedKeys: {
    GEOCODING_API_KEY: "geocoding_api_key", 
    DIRECTIONS_API_KEY: "directions_api_key", // these are null by default.
    PLACES_API_KEY: "places_api_key"
  },
  // Set the max age (in ms) of an item in the cache before it is pruned.
  // Default max age is 5 minutes. Min age is 1 minute and max is 30 days.
  cacheMaxItemAges: {
    reverseGeoCache: 5 * 60000,
    geoCache: 20 * 60000,
    directionsCache: 15 * 60000,
    nearbySearchCache: 300 * 60000,
    placeDetailsCache: 300 * 60000,
  },
  // Set the size of your LRU cache 
  // Default cache sizes are 10 items
  cacheMaxSizes: {
    reverseGeoCache: 20, 
    geoCache: 20,
    directionsCache: 15,
    nearbySearchCache: 10,
    placeDetailsCache: 10
  },
  // Enable printing of a message when a cache is accessed. False by default.
  logCache: true,
  // Use a custom logger function to track cache hits. Defaults to `console.log`.
  logger: console.info
});

(async () => {
  try {

    const geoWrapper = await mapwrap.geocode("Staples Center, LA");
    console.log(geoWrapper.getAllAddresses()); // return an array of objects
    console.log(geoWrapper.getTopAddress()); // return first address in its object
    console.log(geoWrapper.getTopAddress(true)); // return first address as a string

    const reverseGeoWrapper = await mapwrap.reverseGeocode(33.651021, -117.841550);
    console.log(reverseGeoWrapper.getAllAddresses());
    console.log(reverseGeoWrapper.getTopAddress());
    console.log(reverseGeoWrapper.getTopAddress(true));

    /*
    Allowed modes: https://developers.google.com/maps/documentation/directions/intro#TravelModes
    */
    const directionsParams = {
      origin: "Anaheim", // required
      destination: "Irvine", // required
      mode: "driving",
      altRoutes: true, // false by default. retrieve alternate routes (up to 3, usually)
      avoidFerries: false,
      avoidHighways: false,
      avoidIndoor: false,
      avoidTolls: false,
      units: "imperial" // or "metric"
    };

    const directionsWrapper = await mapwrap.directions(directionsParams);

    /* 
    The methods getStartAddress, getEndAddress, getRoute, and getRouteSteps
    accepts an optional number to specify which route you want to retrieve data from.
    Useful only when altRoutes is set to true in directionsParams.
    Defaults to 0 if not specified.
    */
    console.log(directionsWrapper.getStartAddress(1));
    console.log(directionsWrapper.getEndAddress()); 
    console.log(directionsWrapper.getRoute(1)); 
    console.log(directionsWrapper.getRoutes());
    console.log(directionsWrapper.getRouteSteps());

    const placeSearchWrapper = await mapwrap.nearbySearchPlaces({
      location: {
        lat: 33.651021, 
        lng: -117.841550
      },
      radius: 30000, // radius in meters to search nearby.
      keyword: "starbucks", // optional
      // units: "imperial" As of 2.1.0, No longer needed.
    });

    console.log(placeSearchWrapper.getResults());
    /*
    What is next page token?
    https://developers.google.com/places/web-service/search#PlaceSearchPaging
    */
    console.log(placeSearchWrapper.getNextPageToken());

    /* 
    commented out as Google Maps API requires a small amount of time before the next page token becomes valid.
    https://stackoverflow.com/questions/21265756/paging-on-google-places-api-returns-status-invalid-request

    const placeSearchWrapper2 = await mapwrap.additionalPlaces(placeSearchWrapper.getNextPageToken());
    */

    const placeDetailsObject = await mapwrap.placeDetails("ChIJN1t_tDeuEmsRUsoyG83frY4");

    console.log(placeDetailsObject);

  } catch (err) {
    console.log(err);
  }
  
})();

```

If any bugs are found, please don't hesitate to open an issue on the project repo. I'll do my best to address them.

CHANGELOG.md
## Version 2.1.0
- Made `nearbySearchPlaces` no longer require a unit field in the config. `radius` property in the config now accepts a value from 1 to 50000 in meters.

## Version 2.0.1
- Axios dependency updated due to security issues.

## Version 2.0.0
- Config object schema overhauled to provide more flexibility. **New config schema is partially incompatible with 1.x.x versions.**
  - cache size settings have all been renamed and can now be configured via the `cacheMaxSizes` key in the config, with its value being an object with the following optional settings:
    - `reverseGeoCacheSize` renamed to `reverseGeoCache`
    - `geoCacheSize` renamed to `geoCache`
    - `directionsCacheSize` renamed to `directionsCache`
    - `nearbySearchCacheSize` renamed to `nearbySearchCache`
    - `placeDetailsCacheSize` renamed to `placeDetailsCache`
  - Cache max age configuration implemented. Use the `cacheMaxItemAges` key in the config. Its value is an object with the following optional settings:
    - `reverseGeoCache`
    - `geoCache`
    - `directionsCache`
    - `nearbySearchCache`
    - `placeDetailsCache`
    - Each accepts a value in milliseconds. Defaults to 300000 ms (5 minutes). Accepts a minimum of 1 minute and a max of 30 days. When this age is exceeded by an item in its respective LRU cache, the item will be dropped on the next access attempt.

  - to fix the compatibility issues, if you were using any of the renamed config properties you must rename them and place then within `cacheMaxSizes`. Otherwise, no change is needed.

- added optional `logger` config setting for custom logging support. Defaults to `console.log`. Accepts a function.
- Fixed some incorrect docstrings
- updated typings to reflect new changes

## Version 1.1.0
- Fixed error throwing accessing invalid property
- Added typing support
- Improved usage documentation coverage.

## Version 1.0.x
- Only minor fixes to documentation.

## Version 1.0.0
- Initial Release

lib/MapWrap.js
const axios = require("axios");
const LRUCache = require("./cache/LRUCache");
const ReverseGeoWrapper = require("./wrappers/ReverseGeoWrapper");
const GeoWrapper = require("./wrappers/GeoWrapper");
const DirectionsWrapper = require("./wrappers/DirectionsWrapper");
const PlaceSearchWrapper = require("./wrappers/PlaceSearchWrapper");
const Joi = require("joi");
const nearbySearchModel = require("./models/nearbySearchPlaces");
const directionsModel = require("./models/directions");
const mapwrapConfigModel = require("./models/mapwrapConfig");

/**
 * Wrapper for several Google API services including Directions, Geocoding, and Places with built in LRU caching.
 */
class MapWrap {
  /**
   * constructor to configure MapWrap
   * @constructor
   * @param {Object} config - Object to configure MapWrap instance.
   * @param {String} config.DEFAULT_API_KEY - Google API Key to be used. If a key was restricted to a certain service, you must use `useRestrictedKeys`.
   * 
   * @param {Object} [config.useRestrictedKeys] - Alternative option to DEFAULT_API_KEY.
   * @param {String} [config.useRestrictedKeys.GEOCODING_API_KEY] - API key for Geocoding API services.
   * @param {String} [config.useRestrictedKeys.DIRECTIONS_API_KEY] - API key for Directions API services.
   * @param {String} [config.useRestrictedKeys.PLACES_API_KEY] - API key for Places API services.
   * 
   * @param {Boolean} [config.logCache] - enables logging of messages when an item is retrieved from each LRU cache. `false` by default.
   * @param {Function} [config.logger] - Specify a custom logger function to render any messages the instance will produce. `console.log` bu default.
   * 
   * @param {Object} [config.cacheMaxItemAges] - object containing configuration to set a custom max item age for each cache in ms. Cannot be under 1 minute or over 30 days.
   * @param {Number} [config.cacheMaxItemAges.reverseGeoCache] - specify max item age of reverse geocoding cache in ms. 300000 by default.
   * @param {Number} [config.cacheMaxItemAges.geoCache] - specify max item age of geocoding cache in ms. 300000 by default.
   * @param {Number} [config.cacheMaxItemAges.directionsCache] - specify max item age of directions cache in ms. 300000 by default.
   * @param {Number} [config.cacheMaxItemAges.nearbySearchCache] - specify max item age of nearby search cache in ms. 300000 by default.
   * @param {Number} [config.cacheMaxItemAges.placeDetailsCache] - specify max item age of place details cache in ms. 300000 by default.
   * 
   * @param {Object} [config.cacheMaxSizes] - object containing configuration to set a custom size for each cache.
   * @param {Number} [config.cacheMaxSizes.reverseGeoCache] - specify max size of reverse geocoding cache. 10 by default.
   * @param {Number} [config.cacheMaxSizes.geoCache] - specify max size of geocoding cache. 10 by default.
   * @param {Number} [config.cacheMaxSizes.directionsCache] - specify max size of directions cache. 10 by default.
   * @param {Number} [config.cacheMaxSizes.nearbySearchCache] - specify max size of nearby search cache. 10 by default.
   * @param {Number} [config.cacheMaxSizes.placeDetailsCache] - specify max size of place details cache. 10 by default.
   */
  constructor(config) {
    const { value, error } = Joi.validate(config, mapwrapConfigModel);
    
    if (error) {
      // console.log(error);
      throw new Error(error.message);
    }

    const { 
      DEFAULT_API_KEY, 
      logCache, 
      logger,
      useRestrictedKeys, 
      cacheMaxItemAges,
      cacheMaxSizes
    } = value;

    this._keys = {
      GEOCODING_API_KEY: useRestrictedKeys.GEOCODING_API_KEY || DEFAULT_API_KEY,
      DIRECTIONS_API_KEY: useRestrictedKeys.DIRECTIONS_API_KEY || DEFAULT_API_KEY,
      PLACES_API_KEY: useRestrictedKeys.PLACES_API_KEY || DEFAULT_API_KEY
    };

    this._googleMaps = axios.create({
      baseURL: "https://maps.googleapis.com/maps/api"
    });

    this._reverseGeoCache = new LRUCache(cacheMaxSizes.reverseGeoCache, cacheMaxItemAges.reverseGeoCache);
    this._geoCache = new LRUCache(cacheMaxSizes.geoCache, cacheMaxItemAges.geoCache);
    this._directionsCache = new LRUCache(cacheMaxSizes.directionsCache, cacheMaxItemAges.directionsCache);
    this._nearbySearchCache = new LRUCache(cacheMaxSizes.nearbySearchCache, cacheMaxItemAges.nearbySearchCache);
    this._placeDetailsCache = new LRUCache(cacheMaxSizes.placeDetailsCache, cacheMaxItemAges.placeDetailsCache);
    this._logCache = logCache;
    this._logger = logger;
  }

  /**
   * Returns a GeoWrapper object encapsulating the Google API response.
   * https://developers.google.com/maps/documentation/geocoding/intro
   * @param {String} address - Address to search
   * @returns {Promise<GeoWrapper>} - returns a GeoWrapper instance.
   */
  async geocode(address) {
    if (typeof address !== "string")
      throw new TypeError("`address` must be a string");
    
    const data = this._geoCache.get(address);
    if (!data) {
      try {
        const response = await this._googleMaps.get(`/geocode/json?address=${address.replace(/#/g, '')}&key=${this._keys.GEOCODING_API_KEY}`);
        const initData = new GeoWrapper(response.data);
        this._geoCache.set(address, initData);

        return initData;
      } catch (e) {
        throw new Error(e.message);
      }
    } else {
      if (this._logCache) {
        this._logger("MapWrap.geocode: Cache hit.");
      }
      return data;
    }


  }

  /**
   * Returns a ReverseGeoWrapper object encapsulating the Google API response.
   * https://developers.google.com/maps/documentation/geocoding/intro
   * @param {Number} lat - latitude
   * @param {Number} lng - longitude
   * @returns {Promise<ReverseGeoWrapper>} - returns a ReverseGeoWrapper instance.
   */
  async reverseGeocode(lat, lng) {
    if (typeof lat !== "number" || typeof lng !== "number")
      throw new TypeError("lat and lng must be a Number");

    let data = this._reverseGeoCache.get(`${lat},${lng}`);
    if (!data) {
      try {
      
        const response = await this._googleMaps.get(`/geocode/json?latlng=${lat},${lng}&key=${this._keys.GEOCODING_API_KEY}`);
        const initData = new ReverseGeoWrapper(response.data);
        this._reverseGeoCache.set(`${lat},${lng}`, initData);

        return initData;
      } catch (e) {
        throw new Error(e.message);
      }
    } else {
      if (this._logCache) {
        this._logger("MapWrap.reverseGeocode: Cache hit.");
      }
      return data;
    }
  }

  /**
   * Returns a DirectionsWrapper object encapsulating the Google API response.
   * @param {Object} params - object literal of options.
   * @param {String} params.origin - Address of origin
   * @param {String} params.destination - Address of destination
   * @param {String} [params.mode] - Optional. Refer to Google Direction API docs for valid modes of transport. `"driving"` by default.
   * https://developers.google.com/maps/documentation/directions/intro#TravelModes
   * @param {String} [params.units] -  Optional. Can be `"imperial"` or `"metric"`. `"imperial"` by default.
   * @param {Boolean} [params.altRoutes] - Optional. If `true`, search for alternatives routes. `false` by default.
   * @param {Boolean} [params.avoidFerries] - Optional. If `true`, avoid ferries. `false` by default.
   * @param {Boolean} [params.avoidHighways] - Optional. If `true`, avoid highways. `false` by default.
   * @param {Boolean} [params.avoidIndoor] - Optional. If `true`, avoid indoor. `false` by default.
   * @param {Boolean} [params.avoidTolls] - Optional. If `true`, avoid tolls. `false` by default.
   * @returns {Promise<DirectionsWrapper>} - returns a DirectionWrapper instance.
   */
  async directions(params) {
    const { value, error } = Joi.validate(params, directionsModel);
    
    if (error) {
      // error.details.map(e => console.log(e.message));
      throw new Error(error.message);
    }

    if (value.avoidIndoor && value.mode !== "walking")
      throw new Error("`avoidIndoor` only allowed if mode is `walking`");

    let { origin, destination, altRoutes, avoidFerries, avoidHighways, avoidIndoor, avoidTolls, mode, units } = value;

    let avoidArr = [];
    if (avoidTolls) avoidArr.push("tolls"); 
    if (avoidHighways) avoidArr.push("highways");
    if (avoidFerries) avoidArr.push("ferries");
    if (avoidIndoor) avoidArr.push("indoor");
    const avoidStr = avoidArr.join("|");

    const mapsParams = {
      params: {
        origin: origin.replace(/#/g, ''), // regex eliminates invalid characters when making http request
        destination: destination.replace(/#/g, ''),
        mode: mode || "driving",
        alternatives: altRoutes || false,
        units: units || "imperial",
        ...avoidStr && { avoid: avoidStr },
        key: this._keys.DIRECTIONS_API_KEY
      }
    }
    
    const buildCacheKey = `${origin}&${destination}&${mode || "driving"}&${altRoutes || false}&${units || "imperial"}&${avoidStr || "null"}`;
    let data = this._directionsCache.get(buildCacheKey);
    if (!data) {
      try {
        const response = await this._googleMaps.get("/directions/json", mapsParams);
        // wrap data
        // cache it with buildCacheKey
        const initData = new DirectionsWrapper(response.data);
        this._directionsCache.set(buildCacheKey, initData);

        // return wrapped data
        return initData;
      } catch(e) {
        throw new Error(e.message);
      }
    } else {
      if (this._logCache) {
        this._logger("MapWrap.directions: Cache hit.");
      }
      return data;
    }
  }

  /**
   * Returns a PlaceSearchWrapper object encapsulating the Google API response.
   * @param {Object} params - object literal of options.
   * @param {Object} params.location - Specify coordinate to serve as a center to look from in the format of `{ lat: Number, lng: Number }`
   * @param {Number} params.radius - Radius in meters. Must be no less than 1 and no more than 50000.
   * @param {String} [params.keyword] - Optional. Specify a phrase and look for
   * @param {String} [params.type] - Optional. Filter only for locations matching the specified type string.
   * String type from this list: https://developers.google.com/places/web-service/supported_types
   * @param {Number} [params.minprice] - Optional. Number from -1 to 4 specifying the min price range of the location. `-1` or `undefined` to disable. Must be less than `maxprice`.
   * @param {Number} [params.maxprice] - Optional. Number from -1 to 4 specifying the max price range of the location. `-1` or `undefined` to disable.
   * Info on price ranges: https://developers.google.com/places/web-service/search
   * @returns {Promise<PlaceSearchWrapper>} - returns a PlaceSearchWrapper instance.
   */
  async nearbySearchPlaces(params) {
    const { value, error } = Joi.validate(params, nearbySearchModel);
    if (value.minprice > value.maxprice) {
      throw new Error("minprice excees maxprice");
    }
    if (error) {
      // error.details.map(e => console.log(e.message));
      throw new Error(error.message);
    }

    let { location, radius, keyword, type, minprice, maxprice } = value;

    const placesParams = {
      params: {
        location: `${location.lat},${location.lng}`,
        radius,
        ...keyword && { keyword },
        ...type && { type },
        ...(minprice !== -1 && minprice !== undefined) && { minprice },
        ...(maxprice !== -1 && maxprice !== undefined) && { maxprice },
        key: this._keys.PLACES_API_KEY
      }
    };

    const buildCacheKey = `${location.lat},${location.lng}&${radius}&${keyword || ""}&${type || "null"}&${minprice || -1}&${maxprice || -1}`;
    let data = this._nearbySearchCache.get(buildCacheKey);
    if (!data) {
      try {
        const response = await this._googleMaps.get(`/place/nearbysearch/json`, placesParams);
        // wrap data
        // cache it with buildCacheKey
        const initData = new PlaceSearchWrapper(response.data);
        this._nearbySearchCache.set(buildCacheKey, initData);
        // return wrapped data
        return initData;
      } catch(e) {
        throw new Error(e.message);
      }
    } else {
      if (this._logCache) {
        this._logger("MapWrap.nearbySearchPlaces: Cache hit.");
      }
      return data;
    }
  }

  /**
   * Returns a PlaceSearchWrapper object encapsulating the Google API response given a page token. Does not cache results.
   * @param {String} nextPageToken - Page token acquired from Google API response. https://developers.google.com/places/web-service/search#PlaceSearchPaging
   * @returns {Promise<PlaceSearchWrapper>} - returns a PlaceSearchWrapper instance.
   */
  async additionalPlaces(nextPageToken=null) {
    if (!nextPageToken || typeof nextPageToken !== "string") {
      throw new TypeError("`nextPageToken` must be of type `string`");
    }
    const placesParams = {
      params: {
        pagetoken: nextPageToken,
        key: this._keys.PLACES_API_KEY
      }
    };
    
    try {
      const response = await this._googleMaps.get(`/place/nearbysearch/json`, placesParams);
      return new PlaceSearchWrapper(response.data);

    } catch(e) {
      throw new Error(e.message);
    }
  }

  /**
   * Returns an object containing place details.
   * @param {String} placeId - Place ID acquired from Google API response. https://developers.google.com/places/web-service/details
   * @returns {Promise<Object>} - returns an object containing the resulting object with the place details.
   */
  async placeDetails(placeId) {
    if (!placeId || typeof placeId !== "string") {
      throw new TypeError("`placeId` must be of type `string`");
    }

    let data = this._placeDetailsCache.get(placeId);
    if (!data) {
      try {
        const response = await this._googleMaps.get(`/place/details/json?placeid=${placeId}&key=${this._keys.PLACES_API_KEY}`);
      
        this._placeDetailsCache.set(placeId, response.data);
        return response.data;
      } catch(e) {
        throw new Error(e.message);
      }
    } else {
      if (this._logCache) {
        this._logger("MapWrap.placeDetails: Cache hit.");
      }
      return data;
    }   
  }

}

module.exports = MapWrap;